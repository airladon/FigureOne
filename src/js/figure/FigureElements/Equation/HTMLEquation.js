// @flow

function makeDiv(
  id: string,
  classes: Array<string>,
  text: string,
  indent: number = 0,
) {
  const indentStr = ' '.repeat(indent);
  const idStr = id ? ` id="${id}"` : '';
  const classString = classes ? ` ${classes.join(' ')}` : '';
  let out = `${indentStr}<div${idStr} class="equation_element${classString}">\n`;
  out += `${text}\n`;
  out += `${indentStr}</div>`;
  return out;
}

// Most fundamental Equation Element properties includes element size and
// location, as well as html id and classes.
class HTMLElementProperties {
  id: string;
  classes: Array<string>;

  constructor(id: string = '', classes: string | Array<string> = '') {
    this.id = id;
    if (Array.isArray(classes)) {
      this.classes = classes;
    } else if (classes.length > 0) {
      this.classes = classes.split(' ');
    } else {
      this.classes = [];
    }
  }

  render(indent: number = 0, text: string = '') {
    return makeDiv(
      this.id,
      this.classes,
      text,
      indent,
    );
  }
}

class HTMLElement extends HTMLElementProperties {
  text: string;

  constructor(
    text: string,
    id: string = '',
    classes: string | Array<string> = [],
  ) {
    super(id, classes);
    this.classes.push('equation_text');
    this.text = text;
  }

  render(indent: number = 0) {
    return super.render(indent, `${' '.repeat(indent + 2)}${this.text}`);
  }
}

class HTMLElements extends HTMLElementProperties {
  content: Array<HTMLElement | HTMLElements | HTMLElementProperties>;

  constructor(
    content: Array<HTMLElement | HTMLElements | HTMLElementProperties | null>,
    id: string = '',
    classes: string | Array<string> = [],
  ) {
    super(id, classes);
    const filteredContent = [];
    content.forEach((c) => {
      if (c !== null) {
        filteredContent.push(c);
      }
    });
    this.content = filteredContent;
  }

  render(indent: number = 0) {
    return super.render(indent, this.content.map(c => c.render(indent + 2)).join('\n'));
  }
}

class Fraction extends HTMLElementProperties {
  numerator: HTMLElements;
  denominator: HTMLElements;

  constructor(
    numerator: HTMLElements,
    denominator: HTMLElements,
    id: string = '',
    classes: string | Array<string> = [],
  ) {
    super(id, classes);
    this.classes.push('fraction');
    this.numerator = numerator;
    this.denominator = denominator;
  }

  render(indent: number = 0) {
    const s = ' '.repeat(indent + 2);
    let out = '';
    out += `${s}<div class="numerator">\n`;
    out += this.numerator.render(indent + 4);
    out += `\n${s}</div>\n`;
    out += `${s}<div class="fraction_line"> </div>\n`;
    out += `${s}<div class="denominator">\n`;
    out += this.denominator.render(indent + 4);
    out += `\n${s}</div>`;
    return super.render(indent, out);
  }
}

class SuperSub extends HTMLElementProperties {
  superscript: HTMLElements | null;
  subscript: HTMLElements | null;
  content: HTMLElements;

  constructor(
    content: HTMLElements,
    superscript: HTMLElements | null,
    subscript: HTMLElements | null,
    id: string = '',
    classes: string | Array<string> = '',
  ) {
    super(id, classes);
    this.classes.push('supersub');
    this.superscript = superscript;
    this.subscript = subscript;
    this.content = content;
  }

  render(indent: number = 0) {
    const s = ' '.repeat(indent + 2);
    let out = '';
    out += `${s}<div class="supsub_content element">\n`;
    out += this.content.render(indent + 2);
    out += `\n${s}</div>\n`;
    out += `${s}<div class="super_sub element">\n`;
    out += `${s}<div class="super_sub_super superscript_text element">\n`;
    if (this.superscript !== null) {
      out += this.superscript.render(indent + 4);
    }
    out += `\n${s}</div>\n`;
    out += `${s}<div class="super_sub_sub subscript_text element">\n`;
    if (this.subscript !== null) {
      out += this.subscript.render(indent + 4);
    }
    out += `\n${s}</div>`;
    out += `\n${s}</div>\n`;
    return super.render(indent, out);
  }
}

class Subscript extends SuperSub {
  constructor(
    content: HTMLElements,       // eslint-disable-line no-use-before-define
    subscript: HTMLElements,
    id: string = '',
    classes: string | Array<string> = '',
  ) {
    super(content, null, subscript, id, classes);
    const index = this.classes.indexOf('supersub');
    if (index > -1) {
      this.classes.splice(index, 1);
    }
    this.classes.push('subscript');
  }
}

class Superscript extends SuperSub {
  constructor(
    content: HTMLElements,       // eslint-disable-line no-use-before-define
    superscript: HTMLElements,
    id: string = '',
    classes: string | Array<string> = '',
  ) {
    super(content, superscript, null, id, classes);
    const index = this.classes.indexOf('supersub');
    if (index > -1) {
      this.classes.splice(index, 1);
    }
    this.classes.push('superscript');
  }
}


class Root extends HTMLElementProperties {
  content: HTMLElements;

  constructor(
    content: HTMLElements,     // eslint-disable-line no-use-before-define
    id: string = '',
    classes: string | Array<string> = [],
  ) {
    super(id, classes);
    this.content = content;
  }

  render(indent: number = 0) {
    const s = ' '.repeat(indent + 2);
    let out = '';
    out += `${s}<div class="square_root element">\n`;
    out += `${s}  <div class="equation_element radical element">\n`;
    out += `${s}    &radic;\n`;
    out += `${s}  </div>\n`;
    out += this.content.render(indent + 4);
    out += `\n${s}</div>`;
    return super.render(indent, out);
  }
}

// function contentToE(content: string | HTMLElements): HTMLElements {
//   let c;
//   if (typeof content === 'string') {
//     c = new HTMLElements(content);
//   } else {
//     c = content;
//   }
//   return c;
// }

type EquationInput = Array<HTMLElements | HTMLElementProperties |
                           HTMLElement | string> |
                     HTMLElements | HTMLElementProperties |
                     HTMLElement | string;


function contentToElement(content: EquationInput): HTMLElements {
  if (content instanceof HTMLElements) {
    return content;
  }
  if (content instanceof HTMLElement
      || content instanceof HTMLElementProperties) {
    return new HTMLElements([content]);
  }
  if (typeof content === 'string') {
    return new HTMLElements([new HTMLElement(content)]);
  }

  // Otherwise must be array
  const elementArray: Array<HTMLElement | HTMLElements | HTMLElementProperties | null> = [];

  content.forEach((c) => {
    if (typeof c === 'string') {
      elementArray.push(new HTMLElement(c));
    } else {
      elementArray.push(c);
    }
  });
  return new HTMLElements(elementArray);
}


export default class HTMLEquation extends HTMLElements {
  constructor(id: string = '', classes: string | Array<string> = []) {
    super([], id, classes);
  }

  createEq(content: Array<HTMLElements | HTMLElementProperties | string>) {
    const elements = [];
    content.forEach((c) => {
      if (typeof c === 'string') {
        elements.push(new HTMLElement(c));
      } else {
        elements.push(c);
      }
      this.content = elements;
    });
  }

  htmlElement() {
    const element = document.createElement('div');
    element.setAttribute('id', this.id);
    element.innerHTML = this.render();
    this.classes.forEach((c) => {
      if (c) {
        element.classList.add(c);
      }
    });
    return element;
  }

  // eslint-disable-next-line class-methods-use-this
  el(
    content: 'string',
    id: string = '',
    classes: string | Array<string> = [],
  ) {
    return new HTMLElement(content, id, classes);
  }

  // eslint-disable-next-line class-methods-use-this
  supsub(
    content: EquationInput,
    superscript: EquationInput,
    subscript: EquationInput,
    id: string = '',
    classes: string | Array<string> = [],
  ) {
    return new SuperSub(
      contentToElement(content),
      contentToElement(superscript),
      contentToElement(subscript),
      id,
      classes,
    );
  }

  // eslint-disable-next-line class-methods-use-this
  sub(
    content: EquationInput,
    subscript: EquationInput,
    id: string = '',
    classes: string | Array<string> = [],
  ) {
    return new Subscript(
      contentToElement(content),
      contentToElement(subscript),
      id,
      classes,
    );
  }

  // eslint-disable-next-line class-methods-use-this
  sup(
    content: EquationInput,
    superscript: EquationInput,
    id: string = '',
    classes: string | Array<string> = [],
  ) {
    return new Superscript(
      contentToElement(content),
      contentToElement(superscript),
      id,
      classes,
    );
  }


  // e(
  //   content: string | Array<HTMLElementProperties>,
  //   id: string = '',
  //   classes: string | Array<string> = [],
  // ) {
  //   return new HTMLElements(content, id, classes);
  // }

  // eslint-disable-next-line class-methods-use-this
  frac(
    numerator: EquationInput,
    denominator: EquationInput,
    id: string = '',
    classes: string | Array<string> = [],
  ) {
    return new Fraction(
      contentToElement(numerator),
      contentToElement(denominator),
      id,
      classes,
    );
  }

  // eslint-disable-next-line class-methods-use-this
  sqrt(
    content: EquationInput,
    id: string = '',
    classes: string | Array<string> = [],
  ) {
    return new Root(
      contentToElement(content),
      id,
      classes,
    );
  }
}

