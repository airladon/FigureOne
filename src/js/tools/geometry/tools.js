// @flow
import { Point } from './Point';

function pointsToNumbers2(points: Array<Point>) {
  const list = Array(points.length * 2);
  for (let i = 0; i < points.length; i += 1) {
    list[i * 2] = points[i].x;
    list[i * 2 + 1] = points[i].y;
  }
  return list;
}

function pointsToNumbers(points: Array<Point>, dimension: 2 | 3 = 3) {
  if (dimension === 2) {
    return pointsToNumbers2(points);
  }
  const list = Array(points.length * 3);
  for (let i = 0; i < points.length; i += 1) {
    list[i * 3] = points[i].x;
    list[i * 3 + 1] = points[i].y;
    list[i * 3 + 2] = points[i].z;
  }
  return list;
}

function numbersToPoints2(numbers: Array<number>) {
  const points = Array(numbers.length / 2);
  for (let i = 0; i < points.length; i += 1) {
    points[i] = new Point(numbers[i * 2], numbers[i * 2 + 1]);
  }
  return points;
}

function numbersToPoints(numbers: Array<number>, dimension: 2 | 3 = 3) {
  if (dimension === 2) {
    return numbersToPoints2(numbers);
  }
  const points = Array(numbers.length / 3);
  for (let i = 0; i < points.length; i += 1) {
    points[i] = new Point(
      numbers[i * 3],
      numbers[i * 3 + 1],
      numbers[i * 3 + 2],
    );
  }
  return points;
}

function toNumbers(
  pointsOrNumbers: Array<Point> | Array<Array<number>> | Array<number>,
  dimension: 2 | 3 = 3,
) {
  if (pointsOrNumbers.length === 0) {
    return [];
  }
  if (typeof pointsOrNumbers[0] === 'number') {
    return pointsOrNumbers;
  }
  if (Array.isArray(pointsOrNumbers[0])) {
    const out = [];
    for (let i = 0; i < pointsOrNumbers.length; i += 1) {
      out.push(...pointsOrNumbers);
    }
    return out;
  }
  return pointsToNumbers(pointsOrNumbers, dimension);
}

function toPoints(pointsOrNumbers: Array<Point> | Array<number>, dimension: 2 | 3 = 3) {
  if (pointsOrNumbers.length === 0) {
    return [];
  }
  if (typeof pointsOrNumbers[0] !== 'number') {
    return pointsOrNumbers;
  }
  return numbersToPoints(pointsOrNumbers, dimension);
}

export {
  pointsToNumbers2,
  pointsToNumbers,
  numbersToPoints,
  numbersToPoints2,
  toNumbers,
  toPoints,
};
