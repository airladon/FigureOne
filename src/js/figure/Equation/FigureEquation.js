// @flow
import WebGLInstance from '../webgl/webgl';
import {
  Rect,
} from '../../tools/g2';
import {
  FigureElementCollection,
} from '../Element';
import { joinObjects } from '../../tools/tools';
import DrawContext2D from '../DrawContext2D';
import { Equation } from './Equation';
import type { EQN_Equation } from './Equation';
import type { TypeNavigatorOptions } from '../FigureCollections/EqnNavigator';
import EqnNavigator from '../FigureCollections/EqnNavigator';

export type TypeEquationPhrase =
  string
  | Array<string
    | TypeEquationPhrase
    | {                                 // Fraction and SmallFraction
      numerator: TypeEquationPhrase;
      denominator: TypeEquationPhrase;
      symbol: string;
      scaleModifier?: number;
    }
    | {                                 // Superscript and Subscript
      content: TypeEquationPhrase;
      subscript?: TypeEquationPhrase;
      superscript?: TypeEquationPhrase;
    }
    | {                                 // Strike
      content: TypeEquationPhrase;
      symbol: string;
      inSize?: boolean;
    }
    | {                                 // Annotation
      content: TypeEquationPhrase;
      annotationArray: Array<{
        content: TypeEquationPhrase;
        xPosition?: 'left' | 'right' | 'center';
        yPosition?: 'bottom' | 'top' | 'middle' | 'baseline';
        xAlign?: 'left' | 'right' | 'center';
        yAlign?: 'bottom' | 'top' | 'middle' | 'baseline';
        scale?: number;
      }>;
      inSize?: boolean;
    }
    | {                                 // Bracket
      content: TypeEquationPhrase;
      symbol: string;
      symbol: string;
      space?: number;
    }
    | {                                 // Top and Bottom Bar
      content: TypeEquationPhrase;
      symbol: string;
      space?: number;
      outsideSpace?: number;
    }
    | {                                 // Top and Bottom Comment
      content: TypeEquationPhrase;
      comment: TypeEquationPhrase;
      symbol: string;
      space?: number;
      outsideSpace?: number;
    }>;

export default class FigureEquation {
  webgl: Array<WebGLInstance>;
  draw2D: DrawContext2D;
  limits: Rect;
  shapes: Object;
  animateNextFrame: () => void;

  constructor(
    shapes: Object,
    animateNextFrame: () => void,
  ) {
    this.webgl = shapes.webgl;
    this.draw2D = shapes.draw2D;
    this.limits = shapes.limits;
    this.shapes = shapes;
    this.animateNextFrame = animateNextFrame;
  }

  equation(
    options: EQN_Equation,
  ) {
    const equation = new Equation(this.shapes, options);
    return equation;
  }

  addEquation(
    parent: FigureElementCollection,
    name: string,
    options: EQN_Equation = {},
  ) {
    // $FlowFixMe
    const equation = new Equation(this.shapes, options);
    parent.add(name, equation);
    return equation;
  }

  addNavigator(
    parent: FigureElementCollection,
    name: string,
    options: TypeNavigatorOptions,
  ) {
    let navNameToUse = name;
    const optionsToUse = joinObjects({}, options);
    if (optionsToUse.equation == null) {                // $FlowFixMe
      const equation = this.addEquation(parent, `${name}Eqn`, options);
      optionsToUse.equation = equation;
      navNameToUse = `${name}Nav`;
    } else if (!(optionsToUse.equation instanceof Equation)) {
      // let methodPathToUse;
      let nameToUse;
      // let pathToUse;
      let eqnOptions;
      let elementModsToUse;
      // let addElementsToUse;
      let firstScenario;
      if (Array.isArray(optionsToUse.equation)) {
        [, nameToUse, , eqnOptions,
          elementModsToUse, , firstScenario,
        ] = optionsToUse.equation;
      } else {
        nameToUse = optionsToUse.equation.name;
        // pathToUse = optionsToUse.equation.path;
        eqnOptions = optionsToUse.equation.options;
        // methodPathToUse = optionsToUse.equation.method;
        elementModsToUse = optionsToUse.equation.mods;
        firstScenario = optionsToUse.equation.scenario;
      }

      let equation;
      if (Array.isArray(eqnOptions)) {
        equation = this.addEquation(parent, nameToUse, ...eqnOptions);
      } else {
        equation = this.addEquation(parent, nameToUse, eqnOptions);
      }

      if (elementModsToUse != null && elementModsToUse !== {}) {
        equation.setProperties(elementModsToUse);
      }
      if (firstScenario != null && firstScenario in equation.scenarios) {
        equation.setScenario(firstScenario);
      }
      optionsToUse.equation = equation;
    }
    // $FlowFixMe
    const navigator = new EqnNavigator(
      this.shapes,
      this.animateNextFrame,
      optionsToUse,
    );
    parent.add(navNameToUse, navigator);
    return navigator;
  }
}
