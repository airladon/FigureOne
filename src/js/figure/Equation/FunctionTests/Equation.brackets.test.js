// import {
//   Point,
// } from '../../../../tools/g2';
import {
  round,
} from '../../../tools/math';
import * as tools from '../../../tools/tools';
import makeFigure from '../../../__mocks__/makeFigure';
import { Equation } from '../Equation';

tools.isTouchDevice = jest.fn();

jest.mock('../../Gesture');
jest.mock('../../webgl/webgl');
jest.mock('../../DrawContext2D');

describe('Equation Functions - Brackets', () => {
  let figure;
  let eqn;
  let color1;
  let elements;
  let functions;
  // let forms;
  beforeEach(() => {
    figure = makeFigure();
    color1 = [0.95, 0, 0, 1];
    elements = {
      a: 'a',
      b: 'b',
      lb: {
        symbol: 'bracket', side: 'left', lineWidth: 0.01, endLength: 0.03,
      },
      rb: {
        symbol: 'bracket', side: 'right', lineWidth: 0.01, endLength: 0.03,
      },
    };
    functions = {
      single: () => {
        eqn = new Equation(figure.shapes, {
          color: color1,
        });
        const e = eqn.eqn.functions;
        const brac = e.brac.bind(e);
        eqn.addElements(elements);
        eqn.addForms({
          // Full Object
          0: {
            content: {
              brac: {
                left: 'lb',
                content: 'a',
                right: 'rb',
                inSize: true,
                insideSpace: 0.1,
                outsideSpace: 0.1,
                topSpace: 0.1,
                bottomSpace: 0.1,
              },
            },
          },
          // Method Object
          1: {
            brac: {
              left: 'lb',
              content: 'a',
              right: 'rb',
              inSize: true,
              insideSpace: 0.1,
              outsideSpace: 0.1,
              topSpace: 0.1,
              bottomSpace: 0.1,
            },
          },
          // Method Array
          2: { brac: ['lb', 'a', 'rb', true, 0.1, 0.1, 0.1, 0.1] },
          // Function with Method Array
          3: e.brac(['lb', 'a', 'rb', true, 0.1, 0.1, 0.1, 0.1]),
          // Function with parameters
          // 4: e.brac('a', 'lb', 'rb', true, 0.1, 0.1, 0.1, 0.1),
          // // Bound Function with parameters
          // 5: brac('a', 'lb', 'rb', true, 0.1, 0.1, 0.1, 0.1),
          // Bound Function with Object
          4: brac({
            left: 'lb',
            content: 'a',
            right: 'rb',
            inSize: true,
            insideSpace: 0.1,
            outsideSpace: 0.1,
            topSpace: 0.1,
            bottomSpace: 0.1,
          }),
        });
      },
      parameterSteps: () => {
        eqn = new Equation(figure.shapes, { color: color1 });
        const e = eqn.eqn.functions;
        const brac = e.brac.bind(e);
        eqn.addElements(elements);
        eqn.addForms({
          // without
          //   // Method Object
          base: {
            content: [
              {
                brac: {
                  left: 'lb',
                  content: 'a',
                  right: 'rb',
                  inSize: true,
                  insideSpace: 0.1,
                  outsideSpace: 0.1,
                  topSpace: 0.1,
                  bottomSpace: 0.1,
                  minContentHeight: null,
                  minContentDescent: null,
                  height: null,
                  descent: null,
                },
              },
              'b',
            ],
            scale: 1,
          },
          insideSpace: {
            content: [brac([
              'lb', 'a', 'rb', true, 0.2, 0.1, 0.1, 0.1, null, null, null, null,
            ]), 'b'],
            scale: 1,
          },
          outsideSpace: {
            content: [brac([
              'lb', 'a', 'rb', true, 0.1, 0.2, 0.1, 0.1, null, null, null, null,
            ]), 'b'],
            scale: 1,
          },
          topSpace: {
            content: [brac([
              'lb', 'a', 'rb', true, 0.1, 0.1, 0.2, 0.1, null, null, null, null,
            ]), 'b'],
            scale: 1,
          },
          bottomSpace: {
            content: [brac([
              'lb', 'a', 'rb', true, 0.1, 0.1, 0.1, 0.2, null, null, null, null,
            ]), 'b'],
            scale: 1,
          },
          minContentHeight: {
            content: [brac([
              'lb', 'a', 'rb', true, 0.1, 0.1, 0.1, 0.1, 1, null, null, null,
            ]), 'b'],
            scale: 1,
          },
          minContentDescent: {
            content: [brac([
              'lb', 'a', 'rb', true, 0.1, 0.1, 0.1, 0.1, null, 1, null, null,
            ]), 'b'],
            scale: 1,
          },
          minContentDescentWithSmallMinHeight: {
            content: [brac([
              'lb', 'a', 'rb', true, 0.1, 0.1, 0.1, 0.1, 0.1, 1, null, null,
            ]), 'b'],
            scale: 1,
          },
          minContentDescentWithLargeMinHeight: {
            content: [brac([
              'lb', 'a', 'rb', true, 0.1, 0.1, 0.1, 0.1, 2, 1, null, null,
            ]), 'b'],
            scale: 1,
          },
          forceDecentLessThanActualDescent: {
            content: [brac([
              'lb', 'a', 'rb', true, 0.1, 0.1, 0.1, 0.1, null, null, null, 0.004,
            ]), 'b'],
            scale: 1,
          },
          forceDecentGreaterThanActualDescent: {
            content: [brac([
              'lb', 'a', 'rb', true, 0.1, 0.1, 0.1, 0.1, null, null, null, 0.1,
            ]), 'b'],
            scale: 1,
          },
          forceHeightLessThanActualHeight: {
            content: [brac([
              'lb', 'a', 'rb', true, 0.1, 0.1, 0.1, 0.1, null, null, 0.05, null,
            ]), 'b'],
            scale: 1,
          },
          forceHeightGreaterThanActualHeight: {
            content: [brac([
              'lb', 'a', 'rb', true, 0.1, 0.1, 0.1, 0.1, null, null, 1, null,
            ]), 'b'],
            scale: 1,
          },
          notInSize: {
            content: [brac([
              'lb', 'a', 'rb', false, 0.1, 0.1, 0.1, 0.1, null, null, null, null,
            ]), 'b'],
            scale: 1,
          },
          noLeftBracket: {
            content: [brac([
              '', 'a', 'rb', true, 0.1, 0.1, 0.1, 0.1, null, null, null, null,
            ]), 'b'],
            scale: 1,
          },
        });
        figure.elements = eqn;
      },
    };
  });
  describe('Parameter Steps', () => {
    let baseB;
    let baseL;
    let baseR;
    let baseA;
    let baseLScale;
    beforeEach(() => {
      functions.parameterSteps();
      eqn.showForm('base');
      figure.setFirstTransform();
      baseA = eqn._a.getPosition();
      baseB = eqn._b.getPosition();
      baseL = eqn._lb.getPosition();
      baseR = eqn._rb.getPosition();
      baseLScale = eqn._lb.custom.scale._dup();
    });
    // Note, the letter a has the following bounds:
    //   width: 0.1,
    //   height: 0.103,
    //   descent: -0.008,
    //   ascent: top: 0.095,
    // The letter b has the following bounds:
    //   width: 0.1,
    //   height: 0.148,
    //   descent: -0.008,
    //   ascent: top: 0.14,
    test('noLeftBracket', () => {
      eqn.showForm('noLeftBracket');
      figure.setFirstTransform();
      const newA = eqn._a.getBoundingRect('figure');
      const newR = eqn._rb.getBoundingRect('figure');
      expect(round(newA.left)).toBe(0.1);
      expect(round(newR.left)).toBe(round(newA.right + 0.1));
    });
    test('notInSize', () => {
      eqn.showForm('notInSize');
      const newLScale = eqn._lb.custom.scale._dup();
      const newL = eqn._lb.getPosition();
      const newB = eqn._b.getPosition();
      // 0.1x2 + 0.103 = 0.303
      expect(round(newLScale.y)).toBe(0.303);
      expect(round(newL.y)).toBe(-0.108);
      // 0 - 0.1 - 0.04 = -0.14
      expect(round(newL.x)).toBe(round(-0.1 - eqn._lb.getBoundingRect('figure').width));
      expect(round(newB.x)).toBe(0.1);
    });
    test('forceHeightGreaterThanActualHeight', () => {
      eqn.showForm('forceHeightGreaterThanActualHeight');
      const newLScale = eqn._lb.custom.scale._dup();
      const newL = eqn._lb.getPosition();
      // 1
      expect(round(newLScale.y)).toBe(1);
      expect(round(newL.y)).toBe(-0.108);
    });
    test('forceHeightLessThanActualHeight', () => {
      eqn.showForm('forceHeightLessThanActualHeight');
      const newLScale = eqn._lb.custom.scale._dup();
      const newL = eqn._lb.getPosition();
      // 0.005
      expect(round(newLScale.y)).toBe(0.05);
      expect(round(newL.y)).toBe(-0.108);
    });
    test('forceDecentGreaterThanActualDescent', () => {
      eqn.showForm('forceDecentGreaterThanActualDescent');
      const newLScale = eqn._lb.custom.scale._dup();
      const newL = eqn._lb.getPosition();
      // 0.1 + 0.095 + 0.1 = 0.295
      expect(round(newLScale.y)).toBe(0.295);
      expect(round(newL.y)).toBe(-0.1);
    });
    test('forceDecentLessThanActualDescent', () => {
      eqn.showForm('forceDecentLessThanActualDescent');
      const newLScale = eqn._lb.custom.scale._dup();
      const newL = eqn._lb.getPosition();
      // 0.004 + 0.095 + 0.1 = 0.199
      expect(round(newLScale.y)).toBe(0.199);
      expect(round(newL.y)).toBe(-0.004);
    });
    test('minContentDescentWithLargeMinHeight', () => {
      eqn.showForm('minContentDescentWithLargeMinHeight');
      const newLScale = eqn._lb.custom.scale._dup();
      const newL = eqn._lb.getPosition();
      // 1 + 1 + 0.1x2 = 2.2
      expect(round(newLScale.y)).toBe(2.2);
      expect(round(newL.y)).toBe(-1.1);
    });
    test('minContentDescentWithSmallLargeHeight', () => {
      eqn.showForm('minContentDescentWithSmallMinHeight');
      const newLScale = eqn._lb.custom.scale._dup();
      const newL = eqn._lb.getPosition();
      // 1 + 0.095 + 0.1x2 = 1.295 as minHeight doesn't raise above minDescent
      // more than old ascent does
      expect(round(newLScale.y)).toBe(1.295);
      expect(round(newL.y)).toBe(-1.1);
    });
    test('minContentDescentWithSmallMinHeight', () => {
      eqn.showForm('minContentDescentWithSmallMinHeight');
      const newLScale = eqn._lb.custom.scale._dup();
      const newL = eqn._lb.getPosition();
      // 1 + 0.095 + 0.1x2 = 1.295
      expect(round(newLScale.y)).toBe(1.295);
      expect(round(newL.y)).toBe(-1.1);
    });
    test('minContentDescent', () => {
      eqn.showForm('minContentDescent');
      const newLScale = eqn._lb.custom.scale._dup();
      const newL = eqn._lb.getPosition();
      // 1 + 0.095 + 0.1x2 = 1.295
      expect(round(newLScale.y)).toBe(1.295);
      expect(round(newL.y)).toBe(-1.1);
    });
    test('minContentHeight', () => {
      eqn.showForm('minContentHeight');
      const newLScale = eqn._lb.custom.scale._dup();
      const newL = eqn._lb.getPosition();
      // 1 + 0.1x2 = 1.2
      expect(round(newLScale.y)).toBe(1.2);
      expect(round(newL.y)).toBe(round(baseL.y));
    });
    test('bottomSpace', () => {
      eqn.showForm('bottomSpace');
      const newLScale = eqn._lb.custom.scale._dup();
      const newL = eqn._lb.getPosition();
      expect(round(newLScale.y)).toBe(round(baseLScale.y + 0.1));
      expect(round(newL.y)).toBe(round(baseL.y - 0.1));
    });
    test('topSpace', () => {
      eqn.showForm('topSpace');
      const newLScale = eqn._lb.custom.scale._dup();
      const newL = eqn._lb.getPosition();
      expect(round(newLScale.y)).toBe(round(baseLScale.y + 0.1));
      expect(round(newL.y)).toBe(round(baseL.y));
    });
    test('outsideSpace', () => {
      eqn.showForm('outsideSpace');
      const newB = eqn._b.getPosition();
      const newL = eqn._lb.getPosition();
      const newR = eqn._rb.getPosition();
      const newA = eqn._a.getPosition();
      expect(round(newL.x)).toBe(round(baseL.x + 0.1));
      expect(round(newA.x)).toBe(round(baseA.x + 0.1));
      expect(round(newR.x)).toBe(round(baseR.x + 0.1));
      expect(round(newB.x)).toBe(round(baseB.x + 0.2));
    });
    test('insideSpace', () => {
      eqn.showForm('insideSpace');
      figure.setFirstTransform();
      const newB = eqn._b.getPosition();
      const newL = eqn._lb.getPosition();
      const newR = eqn._rb.getPosition();
      const newA = eqn._a.getPosition();
      expect(round(newL.x)).toBe(round(baseL.x));
      expect(round(newA.x)).toBe(round(baseA.x + 0.1));
      expect(round(newR.x)).toBe(round(baseR.x + 0.2));
      expect(round(newB.x)).toBe(round(baseB.x + 0.2));
    });
  });
  test('Bracket', () => {
    functions.single();
    const elems = [eqn._a, eqn._lb, eqn._rb];
    const formsToTest = ['1', '2', '3', '4'];
    eqn.showForm('0');
    const positions0 = elems.map(elem => round(elem.transform.mat).slice());
    formsToTest.forEach((f) => {
      eqn.showForm(f);
      const positions = elems.map(elem => round(elem.transform.mat).slice());
      expect(positions0).toEqual(positions);
    });

    // Snapshot test on most simple layout
    eqn.showForm('0');
    tools.cleanUIDs(eqn);
    expect(round(eqn._a.transform.mat)).toMatchSnapshot();
    expect(round(eqn._lb.transform.mat)).toMatchSnapshot();
    expect(round(eqn._rb.transform.mat)).toMatchSnapshot();
  });
});
