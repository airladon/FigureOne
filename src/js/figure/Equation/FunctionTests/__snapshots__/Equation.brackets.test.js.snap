// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Equation Functions - Brackets Bracket 1`] = `
Array [
  0.7,
  0,
  0,
  0.16841,
  0,
  0.7,
  0,
  0,
  0,
  0,
  1,
  0,
  0,
  0,
  0,
  1,
]
`;

exports[`Equation Functions - Brackets Bracket 2`] = `
Array [
  0.02841,
  0,
  0,
  0.07,
  0,
  0.2121,
  0,
  -0.0756,
  0,
  0,
  1,
  0,
  0,
  0,
  0,
  1,
]
`;

exports[`Equation Functions - Brackets Bracket 3`] = `
Array [
  0.02841,
  0,
  0,
  0.30841,
  0,
  0.2121,
  0,
  -0.0756,
  0,
  0,
  1,
  0,
  0,
  0,
  0,
  1,
]
`;
